Migrate op to new infra: all_reduce_create_qkv_heads

shutovilyaep opened the pull request:
Migrated the `all_reduce_create_qkv_heads` device operation from the old vector-based structure to the new TMP device operation pattern.

Key changes:
- Created new TMP structure with operation_attributes_t, tensor_args_t, and MeshWorkloadFactory pattern for CCL operation
- Replaced old device operation (all_reduce_create_qkv_heads_op.cpp/hpp) with new files (_device_operation.cpp/hpp, _device_operation_types.hpp)
- Updated program factory to use AdaptedCachedMeshWorkload pattern with SharedVariables for runtime argument overrides
- Moved choose_worker_cores_fuse helper to program factory (only user)
- Updated public API to use ttnn::prim::all_reduce_create_qkv_heads
- Maintained backward compatibility - no changes to public API signature

### Ticket
Link to Github Issue #32703

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [RUNNING...] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/20073108226)
- [RUNNING...] Code analysis https://github.com/tenstorrent/tt-metal/actions/runs/20073110956

shutovilyaep committed:
Migrate op to new infra: all_reduce_create_qkv_heads

Fixes #32703

Migrated the `all_reduce_create_qkv_heads` device operation from the old
vector-based structure to the new TMP device operation pattern.

Key changes:
- Created new TMP structure with operation_attributes_t, tensor_args_t,
  and MeshWorkloadFactory pattern for CCL operation
- Replaced old device operation (all_reduce_create_qkv_heads_op.cpp/hpp)
  with new files (_device_operation.cpp/hpp, _device_operation_types.hpp)
- Updated program factory to use AdaptedCachedMeshWorkload pattern
  with SharedVariables for runtime argument overrides
- Moved choose_worker_cores_fuse helper to program factory (only user)
- Updated public API to use ttnn::prim::all_reduce_create_qkv_heads
- Maintained backward compatibility - no changes to public API signature

aliaksei-sala commented on ttnn/cpp/ttnn/operations/experimental/transformer/all_reduce_create_qkv_heads/device/all_reduce_create_qkv_heads_program_factory.cpp:
```diff
@@ -1,51 +1,202 @@
+    [[maybe_unused]] uint32_t input_shard_num_cores = input_tensor_memory_config.shard_spec()->grid.num_cores();
+    [[maybe_unused]] uint32_t output_shard_num_cores = output_tensor_memory_config.shard_spec()->grid.num_cores();
+
+    log_debug(tt::LogOp, "input_tensor_shape: {}", input_tensor_shape);
```
Any reason for new  log_debug messages ?

aliaksei-sala commented on ttnn/cpp/ttnn/operations/experimental/transformer/all_reduce_create_qkv_heads/device/all_reduce_create_qkv_heads_program_factory.cpp:
```diff
@@ -1,51 +1,202 @@
+    tt::tt_metal::distributed::MeshWorkload mesh_workload;
+    std::unordered_map<ttnn::MeshCoordinateRange, shared_variables_t> shared_variables;
+
+    for (const auto& mesh_coord_range : tensor_coords.ranges()) {
```
Is there a reason not to use `ccl::create_mesh_workload_from_programs` as it used to be ?

aliaksei-sala commented on ttnn/cpp/ttnn/operations/experimental/transformer/all_reduce_create_qkv_heads/device/all_reduce_create_qkv_heads_program_factory.cpp:
```diff
@@ -1,51 +1,202 @@
+
+// Helper function to convert cores to corerangeset
+CoreRangeSet cores_to_corerangeset(const std::vector<CoreCoord>& cores) {
+    std::set<CoreRange> core_ranges;
```
It looks like the previously used `cores_to_corerangeset` implementation differs from the current one: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce_async/device/all_reduce_async_program_minimal_variants.cpp.
I'm not sure whether the order has any impact, but during migration it would be safer to preserve the original behavior.

shutovilyaep replied:
fixed:
- Changed to use `std::vector<CoreRange>` with `reserve` and `push_back` instead of `std::set<CoreRange>` to match the original implementation
- Added std::move
- This way, no copying will happen, checked constructors of CoreRangeSet

shutovilyaep committed:
Fix misleading variable names: store addresses instead of buffer pointers

Changed *_base_addr variables to store actual addresses via ->address()
instead of buffer pointers, making the variable names accurate.

shutovilyaep committed:
fix: cores_to_corerangeset, using std::move of std::vector to avoid copying

shutovilyaep commented on ttnn/cpp/ttnn/operations/experimental/transformer/all_reduce_create_qkv_heads/device/all_reduce_create_qkv_heads_program_factory.cpp:
```diff
@@ -621,67 +783,72 @@ tt::tt_metal::operation::ProgramWithCallbacks all_reduce_create_qkv_heads_minima
+    const auto& k_output = tensor_return_value[2];
+    const auto& v_output = tensor_return_value[3];
+
+    auto* q_base_addr = q_output.buffer();
```
fixed

shutovilyaep committed:
fix: Change return types from std::vector to struct with q, k, v members

- Changed spec_return_value_t and tensor_return_value_t from std::vector to structs
- Added all_reduce as first member (internal implementation detail, was previously at index [0])
- Updated compute_output_specs to return struct with all_reduce, q, k, v
- Updated create_output_tensors to create all tensors and return struct
- Updated program factory to use struct members instead of vector indices
- Preserves original logic: all_reduce tensor still created in create_output_tensors
